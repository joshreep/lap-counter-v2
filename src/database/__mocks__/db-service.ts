import mockRunners from '@/test-utils/mock-data/runners'

export const DBService = {
  getAllRunners: jest.fn().mockResolvedValue(mockRunners),
  incrementRunnerLap: jest.fn().mockImplementation(() => Promise.resolve()),
  upsertRunner: jest.fn().mockImplementation(() => Promise.resolve()),
  deleteRunner: jest.fn().mockImplementation(() => Promise.resolve()),
}
