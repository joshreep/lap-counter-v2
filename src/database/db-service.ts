import {
  collection,
  deleteDoc,
  doc,
  getDocs,
  getFirestore,
  increment,
  onSnapshot,
  orderBy,
  query,
  setDoc,
} from 'firebase/firestore'
import { InputRunnerRow, RunnerRow } from './types'
import app from '@/firebaseConfig'
import { useCallback, useEffect, useState } from 'react'

export enum QueryStatus {
  Idle = 'IDLE',
  Loading = 'LOADING',
  Error = 'ERROR',
}

class DBServiceSingleton {
  private _db: ReturnType<typeof getFirestore>
  private _runnersCollectionRef: ReturnType<typeof collection>
  private _runnersListQuery: ReturnType<typeof query>

  constructor() {
    this._db = getFirestore(app)
    this._runnersCollectionRef = collection(this._db, 'runners')
    this._runnersListQuery = query(this._runnersCollectionRef, orderBy('lapCount', 'desc'))
  }

  get db() {
    return this._db
  }

  get runnersCollectionRef() {
    return this._runnersCollectionRef
  }

  get runnersListQuery() {
    return this._runnersListQuery
  }

  async getAllRunners() {
    const querySnapshot = await getDocs(this.runnersListQuery)
    return querySnapshot.docs.map((doc) => doc.data()) as RunnerRow[]
  }

  async incrementRunnerLap(runnerId: string) {
    try {
      const runnerRef = doc(this.runnersCollectionRef, runnerId)
      await setDoc(runnerRef, { runnerId, lapCount: increment(1) }, { merge: true })
    } catch (error) {
      console.error(error)
      throw new Error(`Something went wrong trying to add a lap to Runner number ${runnerId}.`)
    }
  }

  async upsertRunner(runner: InputRunnerRow) {
    try {
      const runnerRef = doc(this.runnersCollectionRef, runner.runnerId)
      await setDoc(runnerRef, runner, { merge: true })
    } catch (error) {
      console.error(error)
      throw new Error(`Something went wrong trying to update Runner number ${runner.runnerId}.`)
    }
  }

  async deleteRunner(runnerId: string) {
    try {
      const runnerRef = doc(this.runnersCollectionRef, runnerId)
      await deleteDoc(runnerRef)
    } catch (error) {
      console.error(error)
      throw new Error(`Something went wrong trying to delete Runner number ${runnerId}.`)
    }
  }
}

export const DBService = new DBServiceSingleton()

export function useRunners() {
  const [data, setData] = useState<RunnerRow[]>([])
  const [status, setStatus] = useState(QueryStatus.Loading)

  const refreshData = useCallback(async () => {
    setStatus(QueryStatus.Loading)
    const data = await DBService.getAllRunners()
    setStatus(QueryStatus.Idle)
    setData(data)
  }, [])

  useEffect(() => {
    return onSnapshot(DBService.runnersListQuery, ({ docs }) => {
      setStatus(QueryStatus.Idle)
      setData(docs.map((doc) => doc.data() as RunnerRow))
    })
  }, [])

  return { data, refreshData, status }
}
