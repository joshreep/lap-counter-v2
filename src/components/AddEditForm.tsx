'use client'

import { DBService } from '@/database/db-service'
import { InputRunnerRow } from '@/database/types'
import { useRouter } from 'next/navigation'
import { FC, useCallback, useRef, useState } from 'react'
import Button from './Button'
import ButtonGroup from './ButtonGroup'
import SubmitAnimation, { SubmissionState } from './SubmitAnimation'
import InputGroup from './form/InputGroup'

export interface AddEditFormProps {
  mode: 'add' | 'edit'
  params?: Record<string, string | undefined>
}

const AddEditForm: FC<AddEditFormProps> = (props) => {
  const { mode, params } = props
  const nameInputRef = useRef<HTMLInputElement>(null)
  const runnerNumberInputRef = useRef<HTMLInputElement>(null)
  const lapCountInputRef = useRef<HTMLInputElement>(null)

  const router = useRouter()

  const [runnerName, setRunnerName] = useState(params?.name ?? '')
  const [runnerNumber, setRunnerNumber] = useState(params?.runnerId ?? '')
  const [lapCount, setLapCount] = useState(params?.lapCount ? +params.lapCount : 0)
  const [submissionState, setSubmissionState] = useState(SubmissionState.Idle)

  const showLapCount = mode === 'edit'

  const onSubmit = useCallback(async () => {
    nameInputRef.current?.blur()
    runnerNumberInputRef.current?.blur()
    if (showLapCount) lapCountInputRef.current?.blur()
    setSubmissionState(SubmissionState.Pending)

    try {
      const input: InputRunnerRow = {
        name: runnerName,
        runnerId: runnerNumber,
        lapCount: lapCount,
      }
      await DBService.upsertRunner(input)
      setSubmissionState(SubmissionState.Complete)
    } catch (error) {
      console.error(error)
      setSubmissionState(SubmissionState.Error)
    }
  }, [lapCount, runnerName, runnerNumber, showLapCount])

  const onSubmitDelete = useCallback(async () => {
    nameInputRef.current?.blur()
    runnerNumberInputRef.current?.blur()
    if (showLapCount) lapCountInputRef.current?.blur()
    setSubmissionState(SubmissionState.Pending)

    try {
      if (params?.runnerId) await DBService.deleteRunner(params.runnerId)
      setSubmissionState(SubmissionState.Complete)
    } catch (error) {
      console.error(error)
      setSubmissionState(SubmissionState.Error)
    }
  }, [params?.runnerId, showLapCount])

  const confirmDelete = useCallback(() => {
    if (
      confirm(
        'Are you Sure?\nAre you Sure you want to delete this runner?  This action cannot be undone!',
      )
    ) {
      onSubmitDelete()
    }
  }, [onSubmitDelete])

  const onCheckAnimationFinish = useCallback(() => {
    setSubmissionState(SubmissionState.Idle)
    if (mode === 'add') {
      setRunnerName('')
      setRunnerNumber('')
      runnerNumberInputRef.current?.focus()
    } else {
      router.back()
    }
  }, [mode, router])

  return (
    <div className="flex flex-col flex-1 items-center justify-start pt-10 pb-5 px-5">
      <form onSubmit={(e) => e.preventDefault()} className="flex flex-col gap-5 w-full">
        <InputGroup
          data-testid="runnerIdInput"
          disabled={mode === 'edit'}
          enterKeyHint="next"
          id="runnerId"
          label="Runner Number"
          min={0}
          onChange={(e) => setRunnerNumber(e.target.value)}
          onBlur={() => nameInputRef.current?.focus()}
          pattern="\d*"
          ref={runnerNumberInputRef}
          type="number"
          value={runnerNumber}
        />
        <InputGroup
          autoCapitalize="words"
          data-testid="runnerNameInput"
          enterKeyHint="next"
          id="name"
          label="Name"
          onChange={(e) => setRunnerName(e.target.value)}
          onBlur={() => showLapCount && lapCountInputRef.current?.focus()}
          ref={nameInputRef}
          value={runnerName}
        />
        {showLapCount && (
          <InputGroup
            data-testid="lapCountInput"
            enterKeyHint="done"
            id="lapCount"
            label="Lap Count"
            min={0}
            onChange={(e) => setLapCount(+e.target.value)}
            pattern="\d*"
            ref={lapCountInputRef}
            type="number"
            value={lapCount.toString()}
          />
        )}
        <ButtonGroup className="flex-row-reverse">
          <Button onClick={onSubmit} className="bg-primary text-white">
            Submit
          </Button>
          {mode === 'edit' && (
            <Button onClick={confirmDelete} buttonStyle="error">
              Delete
            </Button>
          )}
        </ButtonGroup>
      </form>

      <SubmitAnimation
        submissionState={submissionState}
        onAnimationFinish={onCheckAnimationFinish}
      />
    </div>
  )
}

export default AddEditForm
