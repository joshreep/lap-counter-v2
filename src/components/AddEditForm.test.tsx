import { forEachTheme } from '@/test-utils/ThemeWrapper'
import { act, fireEvent, renderRouter } from 'expo-router/testing-library'
import AddEditForm, { AddEditFormProps } from './AddEditForm'
import { DBService } from '@/database/db-service'
import { Alert } from 'react-native'
import { router } from 'expo-router'

const modes: AddEditFormProps['mode'][] = ['add', 'edit']

jest.mock('@/database/db-service')
jest.useFakeTimers()

forEachTheme((theme, wrapper) => {
  describe(`${theme} theme`, () => {
    modes.forEach((mode) => {
      describe(`${mode} mode`, () => {
        afterEach(() => {
          ;(DBService.upsertRunner as jest.Mock).mockReset()
        })

        test('should match snapshot', () => {
          const container = renderRouter({ index: () => <AddEditForm mode={mode} /> }, { wrapper })
          expect(container.toJSON()).toMatchSnapshot()
        })

        test('should handle inputs and submit editing', async () => {
          const { getByTestId } = renderRouter({ index: () => <AddEditForm mode={mode} /> }, { wrapper })

          if (mode === 'add') {
            act(() => fireEvent.changeText(getByTestId('runnerIdInput'), '1'))
            expect(getByTestId('runnerIdInput')).toHaveDisplayValue('1')
            act(() => fireEvent(getByTestId('runnerIdInput'), 'submitEditing'))
          }
          act(() => fireEvent.changeText(getByTestId('runnerNameInput'), 'Thomas Jefferson'))
          expect(getByTestId('runnerNameInput')).toHaveDisplayValue('Thomas Jefferson')
          await act(async () => fireEvent(getByTestId('runnerNameInput'), 'submitEditing'))
          if (mode === 'add') {
            expect(DBService.upsertRunner).toHaveBeenCalled()
          }
          if (mode === 'edit') {
            expect(DBService.upsertRunner).not.toHaveBeenCalled()
            act(() => fireEvent.changeText(getByTestId('lapCountInput'), 101))
            expect(getByTestId('lapCountInput')).toHaveDisplayValue('101')
            await act(async () => fireEvent(getByTestId('lapCountInput'), 'submitEditing'))
            expect(DBService.upsertRunner).toHaveBeenCalled()
          }
        })

        test('should handle submit', async () => {
          const { getByTestId, getByText } = renderRouter(
            { index: () => <AddEditForm mode={mode} />, originPage: () => null },
            { wrapper, initialUrl: '/originPage' },
          )

          act(() => {
            router.navigate('/')
          })
          act(() => {
            mode === 'add' && fireEvent.changeText(getByTestId('runnerIdInput'), '1')
            fireEvent.changeText(getByTestId('runnerNameInput'), 'George Washington')
            mode === 'edit' && fireEvent.changeText(getByTestId('lapCountInput'), 20)
          })
          await act(async () => {
            fireEvent.press(getByText('Submit'))
          })

          expect(DBService.upsertRunner).toHaveBeenCalled()

          // allow animations to finish
          act(() => {
            jest.runAllTimers()
          })

          if (mode === 'add') {
            expect(getByTestId('runnerIdInput')).toHaveDisplayValue('')
            expect(getByTestId('runnerNameInput')).toHaveDisplayValue('')
          }
        })

        if (mode === 'edit') {
          test('should handle delete', async () => {
            const { getByText } = renderRouter(
              {
                index: () => <AddEditForm mode="edit" />,
                page1: () => null,
                page2: () => null,
              },
              { wrapper, initialUrl: '/page1' },
            )
            act(() => {
              router.navigate('/page2')
              router.navigate({ pathname: '/', params: { name: 'George', runnerId: '17', lapCount: 10 } })
            })

            const alertSpy = jest.spyOn(Alert, 'alert')

            act(() => {
              fireEvent.press(getByText('Delete'))
            })

            expect(alertSpy).toHaveBeenCalled()

            await act(async () => {
              alertSpy.mock.calls[0][2]?.[1].onPress?.()
            })

            expect(DBService.deleteRunner).toHaveBeenCalledWith('17')
          })
        }
      })
    })
  })
})
