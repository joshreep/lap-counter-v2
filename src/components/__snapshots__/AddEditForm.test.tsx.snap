// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dark theme add mode should match snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNCSafeAreaView
    edges={
      {
        "bottom": "additive",
        "left": "additive",
        "right": "additive",
        "top": "additive",
      }
    }
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "#111",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "flex-start",
            "paddingBottom": 20,
            "paddingLeft": 20,
            "paddingRight": 20,
            "paddingTop": 40,
          },
          {
            "paddingBottom": 0,
          },
        ]
      }
    >
      <View
        style={
          {
            "backgroundColor": "transparent",
            "display": "flex",
            "gap": 20,
            "width": "100%",
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "transparent",
              "gap": 4,
            }
          }
        >
          <Text
            disabled={false}
            style={
              {
                "color": "rgb(255, 255, 255)",
                "fontSize": 16,
                "marginLeft": 5,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
              }
            }
          >
            Runner Number
          </Text>
          <TextInput
            clearButtonMode="while-editing"
            keyboardType="number-pad"
            onChangeText={[Function]}
            onSubmitEditing={[Function]}
            returnKeyType="done"
            style={
              {
                "backgroundColor": "rgba(255,255,255, .1)",
                "borderBottomLeftRadius": 4,
                "borderBottomRightRadius": 4,
                "borderTopLeftRadius": 4,
                "borderTopRightRadius": 4,
                "color": "rgb(255, 255, 255)",
                "fontSize": 16,
                "paddingBottom": 15,
                "paddingLeft": 15,
                "paddingRight": 15,
                "paddingTop": 15,
                "width": "100%",
              }
            }
            testID="runnerIdInput"
            value=""
          />
        </View>
        <View
          style={
            {
              "backgroundColor": "transparent",
              "gap": 4,
            }
          }
        >
          <Text
            style={
              {
                "color": "rgb(255, 255, 255)",
                "fontSize": 16,
                "marginLeft": 5,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
              }
            }
          >
            Name
          </Text>
          <TextInput
            autoCapitalize="words"
            clearButtonMode="while-editing"
            onChangeText={[Function]}
            onSubmitEditing={[Function]}
            returnKeyType="next"
            style={
              {
                "backgroundColor": "rgba(255,255,255, .1)",
                "borderBottomLeftRadius": 4,
                "borderBottomRightRadius": 4,
                "borderTopLeftRadius": 4,
                "borderTopRightRadius": 4,
                "color": "rgb(255, 255, 255)",
                "fontSize": 16,
                "paddingBottom": 15,
                "paddingLeft": 15,
                "paddingRight": 15,
                "paddingTop": 15,
                "width": "100%",
              }
            }
            testID="runnerNameInput"
            textContentType="name"
            value=""
          />
        </View>
        <View
          style={
            {
              "backgroundColor": "transparent",
              "flexDirection": "row",
              "justifyContent": "space-around",
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                [
                  {},
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                Submit
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaView>
</RNCSafeAreaProvider>
`;

exports[`dark theme edit mode should match snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNCSafeAreaView
    edges={
      {
        "bottom": "additive",
        "left": "additive",
        "right": "additive",
        "top": "additive",
      }
    }
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "#111",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "flex-start",
            "paddingBottom": 20,
            "paddingLeft": 20,
            "paddingRight": 20,
            "paddingTop": 40,
          },
          {
            "paddingBottom": 0,
          },
        ]
      }
    >
      <View
        style={
          {
            "backgroundColor": "transparent",
            "display": "flex",
            "gap": 20,
            "width": "100%",
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "transparent",
              "gap": 4,
            }
          }
        >
          <Text
            disabled={true}
            style={
              {
                "color": "rgb(127, 127, 127)",
                "fontSize": 16,
                "marginLeft": 5,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
              }
            }
          >
            Runner Number
          </Text>
          <View
            style={
              {
                "backgroundColor": "rgba(255, 255, 255, .05)",
                "borderBottomLeftRadius": 4,
                "borderBottomRightRadius": 4,
                "borderTopLeftRadius": 4,
                "borderTopRightRadius": 4,
                "paddingBottom": 15,
                "paddingLeft": 15,
                "paddingRight": 15,
                "paddingTop": 15,
                "width": "100%",
              }
            }
          >
            <Text
              style={
                {
                  "color": "rgb(127, 127, 127)",
                  "fontSize": 16,
                }
              }
              testID="runnerIdInput"
            />
          </View>
        </View>
        <View
          style={
            {
              "backgroundColor": "transparent",
              "gap": 4,
            }
          }
        >
          <Text
            style={
              {
                "color": "rgb(255, 255, 255)",
                "fontSize": 16,
                "marginLeft": 5,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
              }
            }
          >
            Name
          </Text>
          <TextInput
            autoCapitalize="words"
            clearButtonMode="while-editing"
            onChangeText={[Function]}
            onSubmitEditing={[Function]}
            returnKeyType="next"
            style={
              {
                "backgroundColor": "rgba(255,255,255, .1)",
                "borderBottomLeftRadius": 4,
                "borderBottomRightRadius": 4,
                "borderTopLeftRadius": 4,
                "borderTopRightRadius": 4,
                "color": "rgb(255, 255, 255)",
                "fontSize": 16,
                "paddingBottom": 15,
                "paddingLeft": 15,
                "paddingRight": 15,
                "paddingTop": 15,
                "width": "100%",
              }
            }
            testID="runnerNameInput"
            textContentType="name"
            value=""
          />
        </View>
        <View
          style={
            {
              "backgroundColor": "transparent",
              "gap": 4,
            }
          }
        >
          <Text
            style={
              {
                "color": "rgb(255, 255, 255)",
                "fontSize": 16,
                "marginLeft": 5,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
              }
            }
          >
            Lap Count
          </Text>
          <TextInput
            clearButtonMode="while-editing"
            keyboardType="number-pad"
            onChangeText={[Function]}
            onSubmitEditing={[Function]}
            returnKeyType="done"
            style={
              {
                "backgroundColor": "rgba(255,255,255, .1)",
                "borderBottomLeftRadius": 4,
                "borderBottomRightRadius": 4,
                "borderTopLeftRadius": 4,
                "borderTopRightRadius": 4,
                "color": "rgb(255, 255, 255)",
                "fontSize": 16,
                "paddingBottom": 15,
                "paddingLeft": 15,
                "paddingRight": 15,
                "paddingTop": 15,
                "width": "100%",
              }
            }
            testID="lapCountInput"
            value="0"
          />
        </View>
        <View
          style={
            {
              "backgroundColor": "transparent",
              "flexDirection": "row",
              "justifyContent": "space-around",
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                [
                  {},
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                    {
                      "color": "red",
                    },
                  ]
                }
              >
                Delete
              </Text>
            </View>
          </View>
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                [
                  {},
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                Submit
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaView>
</RNCSafeAreaProvider>
`;

exports[`light theme add mode should match snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNCSafeAreaView
    edges={
      {
        "bottom": "additive",
        "left": "additive",
        "right": "additive",
        "top": "additive",
      }
    }
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "rgb(255, 255, 255)",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "flex-start",
            "paddingBottom": 20,
            "paddingLeft": 20,
            "paddingRight": 20,
            "paddingTop": 40,
          },
          {
            "paddingBottom": 0,
          },
        ]
      }
    >
      <View
        style={
          {
            "backgroundColor": "transparent",
            "display": "flex",
            "gap": 20,
            "width": "100%",
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "transparent",
              "gap": 4,
            }
          }
        >
          <Text
            disabled={false}
            style={
              {
                "color": "#111",
                "fontSize": 16,
                "marginLeft": 5,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
              }
            }
          >
            Runner Number
          </Text>
          <TextInput
            clearButtonMode="while-editing"
            keyboardType="number-pad"
            onChangeText={[Function]}
            onSubmitEditing={[Function]}
            returnKeyType="done"
            style={
              {
                "backgroundColor": "rgba(0, 0, 0, .07)",
                "borderBottomLeftRadius": 4,
                "borderBottomRightRadius": 4,
                "borderTopLeftRadius": 4,
                "borderTopRightRadius": 4,
                "color": "#111",
                "fontSize": 16,
                "paddingBottom": 15,
                "paddingLeft": 15,
                "paddingRight": 15,
                "paddingTop": 15,
                "width": "100%",
              }
            }
            testID="runnerIdInput"
            value=""
          />
        </View>
        <View
          style={
            {
              "backgroundColor": "transparent",
              "gap": 4,
            }
          }
        >
          <Text
            style={
              {
                "color": "#111",
                "fontSize": 16,
                "marginLeft": 5,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
              }
            }
          >
            Name
          </Text>
          <TextInput
            autoCapitalize="words"
            clearButtonMode="while-editing"
            onChangeText={[Function]}
            onSubmitEditing={[Function]}
            returnKeyType="next"
            style={
              {
                "backgroundColor": "rgba(0, 0, 0, .07)",
                "borderBottomLeftRadius": 4,
                "borderBottomRightRadius": 4,
                "borderTopLeftRadius": 4,
                "borderTopRightRadius": 4,
                "color": "#111",
                "fontSize": 16,
                "paddingBottom": 15,
                "paddingLeft": 15,
                "paddingRight": 15,
                "paddingTop": 15,
                "width": "100%",
              }
            }
            testID="runnerNameInput"
            textContentType="name"
            value=""
          />
        </View>
        <View
          style={
            {
              "backgroundColor": "transparent",
              "flexDirection": "row",
              "justifyContent": "space-around",
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                [
                  {},
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                Submit
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaView>
</RNCSafeAreaProvider>
`;

exports[`light theme edit mode should match snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNCSafeAreaView
    edges={
      {
        "bottom": "additive",
        "left": "additive",
        "right": "additive",
        "top": "additive",
      }
    }
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "rgb(255, 255, 255)",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "flex-start",
            "paddingBottom": 20,
            "paddingLeft": 20,
            "paddingRight": 20,
            "paddingTop": 40,
          },
          {
            "paddingBottom": 0,
          },
        ]
      }
    >
      <View
        style={
          {
            "backgroundColor": "transparent",
            "display": "flex",
            "gap": 20,
            "width": "100%",
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "transparent",
              "gap": 4,
            }
          }
        >
          <Text
            disabled={true}
            style={
              {
                "color": "rgb(127, 127, 127)",
                "fontSize": 16,
                "marginLeft": 5,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
              }
            }
          >
            Runner Number
          </Text>
          <View
            style={
              {
                "backgroundColor": "rgba(0, 0, 0, .035)",
                "borderBottomLeftRadius": 4,
                "borderBottomRightRadius": 4,
                "borderTopLeftRadius": 4,
                "borderTopRightRadius": 4,
                "paddingBottom": 15,
                "paddingLeft": 15,
                "paddingRight": 15,
                "paddingTop": 15,
                "width": "100%",
              }
            }
          >
            <Text
              style={
                {
                  "color": "rgb(127, 127, 127)",
                  "fontSize": 16,
                }
              }
              testID="runnerIdInput"
            />
          </View>
        </View>
        <View
          style={
            {
              "backgroundColor": "transparent",
              "gap": 4,
            }
          }
        >
          <Text
            style={
              {
                "color": "#111",
                "fontSize": 16,
                "marginLeft": 5,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
              }
            }
          >
            Name
          </Text>
          <TextInput
            autoCapitalize="words"
            clearButtonMode="while-editing"
            onChangeText={[Function]}
            onSubmitEditing={[Function]}
            returnKeyType="next"
            style={
              {
                "backgroundColor": "rgba(0, 0, 0, .07)",
                "borderBottomLeftRadius": 4,
                "borderBottomRightRadius": 4,
                "borderTopLeftRadius": 4,
                "borderTopRightRadius": 4,
                "color": "#111",
                "fontSize": 16,
                "paddingBottom": 15,
                "paddingLeft": 15,
                "paddingRight": 15,
                "paddingTop": 15,
                "width": "100%",
              }
            }
            testID="runnerNameInput"
            textContentType="name"
            value=""
          />
        </View>
        <View
          style={
            {
              "backgroundColor": "transparent",
              "gap": 4,
            }
          }
        >
          <Text
            style={
              {
                "color": "#111",
                "fontSize": 16,
                "marginLeft": 5,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
              }
            }
          >
            Lap Count
          </Text>
          <TextInput
            clearButtonMode="while-editing"
            keyboardType="number-pad"
            onChangeText={[Function]}
            onSubmitEditing={[Function]}
            returnKeyType="done"
            style={
              {
                "backgroundColor": "rgba(0, 0, 0, .07)",
                "borderBottomLeftRadius": 4,
                "borderBottomRightRadius": 4,
                "borderTopLeftRadius": 4,
                "borderTopRightRadius": 4,
                "color": "#111",
                "fontSize": 16,
                "paddingBottom": 15,
                "paddingLeft": 15,
                "paddingRight": 15,
                "paddingTop": 15,
                "width": "100%",
              }
            }
            testID="lapCountInput"
            value="0"
          />
        </View>
        <View
          style={
            {
              "backgroundColor": "transparent",
              "flexDirection": "row",
              "justifyContent": "space-around",
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                [
                  {},
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                    {
                      "color": "red",
                    },
                  ]
                }
              >
                Delete
              </Text>
            </View>
          </View>
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                [
                  {},
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                Submit
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaView>
</RNCSafeAreaProvider>
`;
